### PR TITLE
[lab-supervised-model-evaluation] Sonia Garcia Fominaya

### PR DESCRIPTION
**Pull request by Fominayasg**

### CODE REVIEW

**Comment by breogann** [comment]
Hola Sonia! 

El lab está perfect. Si quisieras plotear la matriz de confusión con [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.plot_confusion_matrix.html):


```python
 from sklearn.metrics import plot_confusion_matrix
 plot_confusion_matrix(clf, X_test, y_test)  
```

Por lo demás, le he puesto este resumen a Álex, y te lo paso que creo que puede ayudar (espero que no confunda más):

Voy a intentar resumir machine learning supervisado de una forma sobre simplificada: utilizando pandas y funciones:

1. Se separan los datos: Variable dependiente y variables independientes. 
Screen Shot 2021-02-26 at 11 50 34


2. **Test/train split:** se separan datos de entrenamiento y datos de validación (se "parte" el dataset en cuatro)

Screen Shot 2021-02-26 at 11 50 44

3. **fit**: el modelo "busca" una función que aplicada a X_train, devuelva el valor más parecido a y_train. 
    1. Coge el valor de la X, le pasa la función  "f(x) = x + 2". Ya ha encontrado la función/algoritmo/fórmula que mejor se ajusta (**fits**) los datos (concretamente que mejor se ajusta a X_train, que es con lo que hemos entrenado).
    2. X_train & y_train ya NO nos hace falta, porque son las variables que se han utilizado para poder obtener esa fórmula.

    ```python
    def fit (X_train, y_train):
    	'''Entrena con los datos
    Recibe: datos de entrada y datos de salida con los que comparar
    Devuelve: un algoritmo/función/fórmula'''
    	return una_función
    ```

4. **predict**: ya con la fórmula encontrada, se genera una nueva columna en el dataframe donde hace:

    ```python
    df['y_predict'] = df['y_test'].apply(lambda x: x + 2)
    ```

    entonces ahora tenemos un valor predicho por cada valor de y_test. 

    ```python
    def predict (y_test, fit):
    '''Aplica la función que ha generado fit a un dato para generar uno nuevo. 
    Recibe: el dato & la función
    Devuelve: un valor predicho'''
    	return un_valor_predicho
    ```
Screen Shot 2021-02-26 at 11 52 51


5. **error/accuracy:** partimos con el dataframe con una y_test, y ahora tenemos una columna nueva que es y_pred. Entonces ahora podemos calcular cuánto se equivocan o no la y de partida y la y que hemos predicho. Se comparan los dos valores.

    ```python
    def (y_test, y_pred):
    '''Calcula cómo de bien funciona el modelo para datos con los que no ha entrenado
    Recibe: y original & y predicha con la función predict'''
    Devuelve: error, accuracy
    	return error, accuracy
    ```

Screen Shot 2021-02-26 at 11 53 03


Resumiendo, es una especie de pipeline que funciona de esta forma: train/test split (para después poder validar con datos con los que no se entrenan y evitar over-fitting) ➡️ fit (encontrar fórmula con datos de entrenamiento) ➡️ predict (generar datos predichos sobre datos no entrenados) ➡️ error (comparando datos no entrenados con su predicción). 

Entonces: **X_train** y **y_train** una vez hecho el fit, no lo utilizamos. El ejercicio te pide que aun así utilices esos valores para sacar esta conclusión: cuando se predice y se valora un modelo con los datos que se han utilizado para entrenar, las métricas serán mejores porque hay overfitting. Es decir, que predecirá súper bien ese dataset, pero no sabe cómo predecirá datos nuevos. Y esos datos nuevos serían el X_test y el y_test.
